File: android-app/app/src/main/java/com/inputblocker/app/OverlayService.java
package com.inputblocker.app;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PixelFormat;
import android.os.Build;
import android.os.IBinder;
import android.util.Log;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.List;

public class OverlayService extends Service {

    private static final String TAG = "InputBlocker-Overlay";
    private static final String CHANNEL_ID = "InputBlockerOverlay";
    private static final int NOTIFICATION_ID = 1001;

    private WindowManager windowManager;
    private TouchBlockView touchBlockView;
    private List<Region> regions = new ArrayList<>();
    private boolean isEnabled = true;
    private boolean forceSafeMode = false;

    private BroadcastReceiver configReceiver;

    @Override
    public void onCreate() {
        super.onCreate();
        
        Log.i(TAG, "OverlayService starting...");
        
        createNotificationChannel();
        startForeground(NOTIFICATION_ID, createNotification());
        
        windowManager = (WindowManager) getSystemService(WINDOW_SERVICE);
        
        loadConfig();
        createOverlayView();
        registerConfigReceiver();
        
        Log.i(TAG, "OverlayService ready - regions: " + regions.size() + ", enabled: " + isEnabled);
    }

    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(
                    CHANNEL_ID,
                    "InputBlocker Active",
                    NotificationManager.IMPORTANCE_LOW
            );
            channel.setDescription("Shows when InputBlocker is blocking touches");
            channel.setShowBadge(false);
            
            NotificationManager manager = getSystemService(NotificationManager.class);
            if (manager != null) {
                manager.createNotificationChannel(channel);
            }
        }
    }

    private Notification createNotification() {
        Intent notificationIntent = new Intent(this, MainActivity.class);
        PendingIntent pendingIntent = PendingIntent.getActivity(
                this, 0, notificationIntent, 
                PendingIntent.FLAG_IMMUTABLE | PendingIntent.FLAG_UPDATE_CURRENT
        );

        String statusText;
        if (forceSafeMode) {
            statusText = "Safe Mode (Disabled) - Reset to enable";
        } else if (isEnabled) {
            statusText = "Blocking " + regions.size() + " region(s)";
        } else {
            statusText = "Blocking disabled";
        }

        return new NotificationCompat.Builder(this, CHANNEL_ID)
                .setContentTitle("InputBlocker")
                .setContentText(statusText)
                .setSmallIcon(android.R.drawable.ic_dialog_info)
                .setContentIntent(pendingIntent)
                .setOngoing(true)
                .setPriority(NotificationCompat.PRIORITY_LOW)
                .build();
    }

    private void loadConfig() {
        regions.clear();
        
        File configFile = new File(InputBlockerServiceManager.getConfigFile(this));
        if (!configFile.exists()) {
            Log.w(TAG, "Config file not found");
            return;
        }

        try (BufferedReader reader = new BufferedReader(new FileReader(configFile))) {
            String line;
            while ((line = reader.readLine()) != null) {
                line = line.trim();
                if (line.isEmpty() || line.startsWith("#")) continue;

                if (line.startsWith("enabled=")) {
                    isEnabled = line.substring(8).equals("1");
                    Log.i(TAG, "Config enabled: " + isEnabled);
                    continue;
                }

                if (line.startsWith("force_safe_mode=")) {
                    forceSafeMode = line.substring(15).equals("1");
                    if (forceSafeMode) {
                        isEnabled = false;
                        Log.i(TAG, "Safe mode is active - blocking disabled");
                    }
                    continue;
                }

                String[] parts = line.split(",");
                if (parts.length == 4) {
                    try {
                        Region region = new Region();
                        region.x1 = Integer.parseInt(parts[0].trim());
                        region.y1 = Integer.parseInt(parts[1].trim());
                        region.x2 = Integer.parseInt(parts[2].trim());
                        region.y2 = Integer.parseInt(parts[3].trim());
                        regions.add(region);
                    } catch (NumberFormatException e) {
                        Log.w(TAG, "Invalid region line: " + line);
                    }
                }
            }
        } catch (Exception e) {
            Log.e(TAG, "Error loading config", e);
        }
        
        Log.i(TAG, "Loaded " + regions.size() + " regions");
    }

    private void createOverlayView() {
        if (windowManager == null) {
            Log.e(TAG, "WindowManager is null");
            return;
        }

        touchBlockView = new TouchBlockView(this);
        
        WindowManager.LayoutParams params = new WindowManager.LayoutParams(
                WindowManager.LayoutParams.MATCH_PARENT,
                WindowManager.LayoutParams.MATCH_PARENT,
                WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY,
                WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE |
                WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL |
                WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN |
                WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS,
                PixelFormat.TRANSLUCENT
        );

        params.gravity = Gravity.TOP | Gravity.START;
        
        touchBlockView.setRegions(regions);
        touchBlockView.setEnabled(isEnabled && !forceSafeMode);

        try {
            windowManager.addView(touchBlockView, params);
            Log.i(TAG, "Overlay view added");
        } catch (Exception e) {
            Log.e(TAG, "Failed to add overlay view", e);
        }
    }

    private void registerConfigReceiver() {
        configReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                String action = intent.getAction();
                Log.i(TAG, "Received broadcast: " + action);
                
                if ("com.inputblocker.RELOAD".equals(action)) {
                    reloadConfig();
                } else if ("com.inputblocker.DISABLE".equals(action)) {
                    disableBlocking();
                } else if ("com.inputblocker.ENABLE".equals(action)) {
                    boolean forceSafe = intent.getBooleanExtra("force_safe", true);
                    enableBlocking(forceSafe);
                }
            }
        };

        IntentFilter filter = new IntentFilter();
        filter.addAction("com.inputblocker.RELOAD");
        filter.addAction("com.inputblocker.DISABLE");
        filter.addAction("com.inputblocker.ENABLE");
        
        try {
            registerReceiver(configReceiver, filter);
        } catch (Exception e) {
            Log.e(TAG, "Failed to register receiver", e);
        }
    }

    private void reloadConfig() {
        loadConfig();
        
        if (touchBlockView != null) {
            touchBlockView.setRegions(regions);
            touchBlockView.setEnabled(isEnabled && !forceSafeMode);
        }
        
        NotificationManager nm = getSystemService(NotificationManager.class);
        if (nm != null) {
            nm.notify(NOTIFICATION_ID, createNotification());
        }
        
        Log.i(TAG, "Config reloaded");
    }

    private void disableBlocking() {
        isEnabled = false;
        
        if (touchBlockView != null) {
            touchBlockView.setEnabled(false);
        }
        
        NotificationManager nm = getSystemService(NotificationManager.class);
        if (nm != null) {
            nm.notify(NOTIFICATION_ID, createNotification());
        }
        
        Toast.makeText(this, "InputBlocker disabled", Toast.LENGTH_SHORT).show();
        Log.i(TAG, "Blocking disabled");
    }

    private void enableBlocking(boolean clearSafeMode) {
        if (clearSafeMode) {
            forceSafeMode = false;
            clearSafeModeFlag();
        }
        
        isEnabled = true;
        
        if (touchBlockView != null) {
            touchBlockView.setEnabled(true);
        }
        
        NotificationManager nm = getSystemService(NotificationManager.class);
        if (nm != null) {
            nm.notify(NOTIFICATION_ID, createNotification());
        }
        
        Toast.makeText(this, "InputBlocker enabled", Toast.LENGTH_SHORT).show();
        Log.i(TAG, "Blocking enabled");
    }

    private void clearSafeModeFlag() {
        try {
            File configFile = new File(InputBlockerServiceManager.getConfigFile(this));
            if (configFile.exists()) {
                StringBuilder content = new StringBuilder();
                BufferedReader reader = new BufferedReader(new FileReader(configFile));
                
                String line;
                while ((line = reader.readLine()) != null) {
                    if (!line.startsWith("force_safe_mode=")) {
                        content.append(line).append("\n");
                    }
                }
                reader.close();
                
                FileWriter writer = new FileWriter(configFile);
                writer.write(content.toString());
                writer.close();
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to clear safe mode flag", e);
        }
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.i(TAG, "onStartCommand");
        
        if (intent != null) {
            String action = intent.getAction();
            if ("RELOAD".equals(action)) {
                reloadConfig();
            } else if ("DISABLE".equals(action)) {
                disableBlocking();
            } else if ("ENABLE".equals(action)) {
                enableBlocking(true);
            }
        }
        
        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        
        if (configReceiver != null) {
            try {
                unregisterReceiver(configReceiver);
            } catch (Exception ignored) {}
        }
        
        if (touchBlockView != null && windowManager != null) {
            try {
                windowManager.removeView(touchBlockView);
            } catch (Exception ignored) {}
        }
        
        Log.i(TAG, "OverlayService destroyed");
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    public static class Region {
        public int x1, y1, x2, y2;
    }

    public class TouchBlockView extends View {
        private Paint blockPaint;
        private Paint borderPaint;
        private Paint textPaint;
        private Paint safeModePaint;
        private List<Region> regions = new ArrayList<>();
        private boolean enabled = true;

        public TouchBlockView(Context context) {
            super(context);
            init();
        }

        private void init() {
            blockPaint = new Paint();
            blockPaint.setColor(Color.parseColor("#1A00FF00"));
            blockPaint.setStyle(Paint.Style.FILL);

            borderPaint = new Paint();
            borderPaint.setColor(Color.parseColor("#00FF00"));
            borderPaint.setStyle(Paint.Style.STROKE);
            borderPaint.setStrokeWidth(4);

            textPaint = new Paint();
            textPaint.setColor(Color.parseColor("#00FF00"));
            textPaint.setTextSize(36);
            textPaint.setTextAlign(Paint.Align.LEFT);
            textPaint.setFakeBoldText(true);

            safeModePaint = new Paint();
            safeModePaint.setColor(Color.parseColor("#FFFF00"));
            safeModePaint.setTextSize(48);
            safeModePaint.setTextAlign(Paint.Align.CENTER);
            safeModePaint.setFakeBoldText(true);
        }

        public void setRegions(List<Region> regions) {
            this.regions = new ArrayList<>(regions);
            invalidate();
        }

        public void setEnabled(boolean enabled) {
            this.enabled = enabled;
            invalidate();
        }

        @Override
        protected void onDraw(Canvas canvas) {
            super.onDraw(canvas);

            if (!enabled) {
                canvas.drawColor(Color.TRANSPARENT);
                
                if (forceSafeMode) {
                    canvas.drawText("SAFE MODE", canvas.getWidth() / 2f, 100, safeModePaint);
                    canvas.drawText("(Blocking Disabled)", canvas.getWidth() / 2f, 150, safeModePaint);
                }
                return;
            }

            for (Region region : regions) {
                android.graphics.RectF rect = new android.graphics.RectF(
                        region.x1, region.y1, region.x2, region.y2
                );
                
                canvas.drawRect(rect, blockPaint);
                canvas.drawRect(rect, borderPaint);
            }

            if (!regions.isEmpty()) {
                float padding = 10;
                canvas.drawText("BLOCKED: " + regions.size() + " region(s)", 
                        padding, 50, textPaint);
            } else {
                canvas.drawText("No regions configured", 10, 50, textPaint);
            }
        }

        @Override
        public boolean onTouchEvent(MotionEvent event) {
            if (!enabled || regions.isEmpty()) {
                return false;
            }

            float x = event.getX();
            float y = event.getY();

            if (event.getAction() == MotionEvent.ACTION_DOWN ||
                event.getAction() == MotionEvent.ACTION_MOVE) {
                
                for (Region region : regions) {
                    if (x >= region.x1 && x <= region.x2 &&
                        y >= region.y1 && y <= region.y2) {
                        
                        Log.d(TAG, "Blocked touch at (" + x + "," + y + ")");
                        return true;
                    }
                }
            }

            return false;
        }
    }
}

File: android-app/app/src/main/java/com/inputblocker/app/OverlayService.kt
package com.inputblocker.app

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.PixelFormat
import android.os.Build
import android.os.IBinder
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import android.util.Log
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.view.WindowManager
import android.widget.Toast
import androidx.annotation.Nullable
import androidx.core.app.NotificationCompat
import com.inputblocker.shared.Region
import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.lang.ref.WeakReference
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.atomic.AtomicInteger

class OverlayService : Service() {

    companion object {
        private const val TAG = "InputBlocker-Overlay"
        private const val CHANNEL_ID = "InputBlockerOverlay"
        private const val NOTIFICATION_ID = 1001

        /** Pause durations (ms) for notification action buttons */
        private const val PAUSE_5MIN_MS = 5 * 60 * 1000L
        private const val PAUSE_30MIN_MS = 30 * 60 * 1000L

        /** Minimum gap (ms) between block log entries */
        private const val RATE_LIMIT_MS = 300L

        private const val ACTION_PAUSE_5MIN = "com.inputblocker.overlay.PAUSE_5MIN"
        private const val ACTION_PAUSE_30MIN = "com.inputblocker.overlay.PAUSE_30MIN"
        private const val ACTION_RESUME = "com.inputblocker.overlay.RESUME"

        private val globalBlockLog = ConcurrentLinkedQueue<BlockLogActivity.BlockEntry>()

        fun getRecentBlocks(): List<BlockLogActivity.BlockEntry> {
            return globalBlockLog.toList().reversed()
        }

        fun addBlockEntry(entry: BlockLogActivity.BlockEntry) {
            globalBlockLog.add(entry)
            if (globalBlockLog.size > 50) {
                globalBlockLog.poll()
            }
        }
    }

    private var windowManager: WindowManager? = null
    private var touchBlockView: TouchBlockView? = null
    private val regions = CopyOnWriteArrayList<Region>()
    @Volatile private var isEnabled = true
    @Volatile private var forceSafeMode = false
    @Volatile private var paused = false
    private var pauseExpirationTime = 0L
    private var currentProfile = "default"
    private var lastForegroundApp: String? = null
    private var blockingExpirationTime = 0L

    private var configReceiver: BroadcastReceiver? = null
    private var configObserver: ConfigFileObserver? = null
    private var isRunning = true
    private var overlayParams: WindowManager.LayoutParams? = null

    // ── Rate limiting & block counter ───────────────────────────────
    private var lastBlockLogTime = 0L
    private val blockCount = AtomicInteger(0)

    // ── Notification action receiver ────────────────────────────────
    private val notificationReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            when (intent.action) {
                ACTION_PAUSE_5MIN -> pauseBlocking(PAUSE_5MIN_MS)
                ACTION_PAUSE_30MIN -> pauseBlocking(PAUSE_30MIN_MS)
                ACTION_RESUME -> resumeBlocking()
            }
        }
    }

    override fun onCreate() {
        super.onCreate()
        Log.i(TAG, "OverlayService starting...")
        isRunning = true
        createNotificationChannel()

        // Register notification action receiver
        val filter = IntentFilter().apply {
            addAction(ACTION_PAUSE_5MIN)
            addAction(ACTION_PAUSE_30MIN)
            addAction(ACTION_RESUME)
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            registerReceiver(notificationReceiver, filter, RECEIVER_NOT_EXPORTED)
        } else {
            @Suppress("UnspecifiedRegisterReceiverFlag")
            registerReceiver(notificationReceiver, filter)
        }

        startForeground(NOTIFICATION_ID, createNotification())
        windowManager = getSystemService(WINDOW_SERVICE) as WindowManager
        loadConfig()
        createOverlayView()
        registerConfigReceiver()
        startConfigWatching()
        startPollLoop()
    }

    // ═══════════════════════════════════════════════════════════════
    //  Config Monitoring (FileObserver + polling fallback)
    // ═══════════════════════════════════════════════════════════════

    private fun startConfigWatching() {
        configObserver?.stop()
        val configDir = File(InputBlockerServiceManager.getConfigDir(this))
        if (!configDir.exists()) configDir.mkdirs()

        val observer = ConfigFileObserver(
            configDir.absolutePath,
            { reloadConfig() }
        )
        observer.start()
        configObserver = observer
    }

    private fun startPollLoop() {
        Thread {
            while (isRunning) {
                try {
                    Thread.sleep(2000)
                    if (!isRunning) break

                    // Foreground app detection
                    val fgApp = getForegroundPackage()
                    if (fgApp != null && fgApp != lastForegroundApp) {
                        lastForegroundApp = fgApp
                        val profileFile = File(InputBlockerServiceManager.getConfigFile(this, fgApp))
                        if (profileFile.exists()) {
                            currentProfile = fgApp
                            reloadConfig()
                        } else if (currentProfile != "default") {
                            currentProfile = "default"
                            reloadConfig()
                        }
                    }

                    // Auto-expire blocking session
                    if (isEnabled && blockingExpirationTime > 0 &&
                        System.currentTimeMillis() > blockingExpirationTime
                    ) {
                        disableBlocking()
                        runOnUiThread {
                            Toast.makeText(this@OverlayService, "Blocking session expired", Toast.LENGTH_LONG).show()
                        }
                    }

                    // Auto-resume from pause
                    if (paused && System.currentTimeMillis() > pauseExpirationTime) {
                        paused = false
                        runOnUiThread {
                            updateOverlayState()
                            Toast.makeText(this@OverlayService, "Blocking resumed", Toast.LENGTH_SHORT).show()
                        }
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "Poll loop error", e)
                }
            }
        }.apply { isDaemon = true }.start()
    }

    // ═══════════════════════════════════════════════════════════════
    //  Pause / Resume
    // ═══════════════════════════════════════════════════════════════

    private fun pauseBlocking(durationMs: Long) {
        paused = true
        pauseExpirationTime = System.currentTimeMillis() + durationMs
        runOnUiThread {
            updateOverlayState()
            val mins = durationMs / 60000
            Toast.makeText(this, "Blocking paused for $mins min", Toast.LENGTH_SHORT).show()
            val nm = getSystemService(NotificationManager::class.java)
            nm?.notify(NOTIFICATION_ID, createNotification())
        }
    }

    private fun resumeBlocking() {
        paused = false
        pauseExpirationTime = 0L
        runOnUiThread {
            updateOverlayState()
            Toast.makeText(this, "Blocking resumed", Toast.LENGTH_SHORT).show()
            val nm = getSystemService(NotificationManager::class.java)
            nm?.notify(NOTIFICATION_ID, createNotification())
        }
    }

    // ═══════════════════════════════════════════════════════════════
    //  Legacy helper
    // ═══════════════════════════════════════════════════════════════

    private fun isLsposedModeFromConfig(): Boolean {
        val configFile = File(InputBlockerServiceManager.getConfigFile(this, "default"))
        if (!configFile.exists()) return false
        return try {
            configFile.readLines().any { it.trim().startsWith("lsposed_mode=1") }
        } catch (e: Exception) {
            false
        }
    }

    private fun getForegroundPackage(): String? {
        val output = InputBlockerServiceManager.runRootCommand("dumpsys activity activities | grep mResumedActivity")
        if (output.isBlank()) return null
        val match = Regex(" ([a-zA-Z0-9._]+)/").find(output)
        return match?.groupValues?.get(1)
    }

    // ═══════════════════════════════════════════════════════════════
    //  Notification
    // ═══════════════════════════════════════════════════════════════

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                CHANNEL_ID, "InputBlocker Active", NotificationManager.IMPORTANCE_LOW
            )
            val manager = getSystemService(NotificationManager::class.java)
            manager?.createNotificationChannel(channel)
        }
    }

    private fun createNotification(): Notification {
        val notificationIntent = Intent(this, MainActivity::class.java)
        val pendingIntent = PendingIntent.getActivity(
            this, 0, notificationIntent, PendingIntent.FLAG_IMMUTABLE
        )

        val toggleIntent = Intent(this, NotificationReceiver::class.java).apply {
            action = NotificationReceiver.ACTION_TOGGLE_BLOCKING
        }
        val togglePendingIntent = PendingIntent.getBroadcast(
            this, 1, toggleIntent, PendingIntent.FLAG_IMMUTABLE
        )

        val safeIntent = Intent(this, NotificationReceiver::class.java).apply {
            action = NotificationReceiver.ACTION_SAFE_MODE
        }
        val safePendingIntent = PendingIntent.getBroadcast(
            this, 2, safeIntent, PendingIntent.FLAG_IMMUTABLE
        )

        val statusText = buildString {
            when {
                forceSafeMode -> append("Safe Mode")
                paused -> append("Paused")
                isEnabled -> append("Active — ${regions.size} zone(s)")
                else -> append("Disabled")
            }
            if (blockCount.get() > 0) {
                append(" [${blockCount.get()} blocked]")
            }
        }

        val builder = NotificationCompat.Builder(this, CHANNEL_ID)
            .setContentTitle("InputBlocker")
            .setContentText(statusText)
            .setSmallIcon(android.R.drawable.ic_lock_idle_lock)
            .setContentIntent(pendingIntent)
            .setOngoing(true)
            .addAction(
                android.R.drawable.ic_menu_close_clear_cancel,
                "Toggle", togglePendingIntent
            )
            .addAction(
                android.R.drawable.ic_menu_view,
                "Safe Mode", safePendingIntent
            )

        // Add pause/resume actions when blocking is active
        if (!paused && isEnabled && !forceSafeMode) {
            builder.addAction(
                android.R.drawable.ic_media_pause,
                "Pause 5m",
                PendingIntent.getBroadcast(
                    this, 3,
                    Intent(ACTION_PAUSE_5MIN).setPackage(packageName),
                    PendingIntent.FLAG_IMMUTABLE
                )
            )
            builder.addAction(
                android.R.drawable.ic_media_pause,
                "Pause 30m",
                PendingIntent.getBroadcast(
                    this, 4,
                    Intent(ACTION_PAUSE_30MIN).setPackage(packageName),
                    PendingIntent.FLAG_IMMUTABLE
                )
            )
        } else if (paused) {
            builder.addAction(
                android.R.drawable.ic_media_play,
                "Resume",
                PendingIntent.getBroadcast(
                    this, 5,
                    Intent(ACTION_RESUME).setPackage(packageName),
                    PendingIntent.FLAG_IMMUTABLE
                )
            )
        }

        return builder.build()
    }

    // ═══════════════════════════════════════════════════════════════
    //  Config Loading
    // ═══════════════════════════════════════════════════════════════

    private fun loadConfig() {
        val oldRegions = regions.toList()
        regions.clear()
        val configFile = File(InputBlockerServiceManager.getConfigFile(this, currentProfile))
        if (!configFile.exists()) {
            if (currentProfile != "default") {
                currentProfile = "default"
                loadConfig()
            }
            return
        }

        try {
            BufferedReader(FileReader(configFile)).use { reader ->
                reader.lineSequence().forEach { line ->
                    val trimmed = line.trim()
                    when {
                        trimmed.startsWith("enabled=") -> isEnabled = trimmed.substring(8) == "1"
                        trimmed.startsWith("force_safe_mode=") -> {
                            forceSafeMode = trimmed.substring(15) == "1"
                            if (forceSafeMode) isEnabled = false
                        }
                        trimmed.isNotEmpty() && !trimmed.startsWith("#") -> {
                            Region.fromString(trimmed)?.let { regions.add(it) }
                        }
                    }
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error loading config", e)
        }

        // Reset block count when regions actually change
        if (regions.toList() != oldRegions) {
            blockCount.set(0)
        }
    }

    private fun reloadConfig() {
        loadConfig()
        runOnUiThread { updateOverlayState() }
        val nm = getSystemService(NotificationManager::class.java)
        nm?.notify(NOTIFICATION_ID, createNotification())
    }

    private fun emergencyReset() {
        InputBlockerServiceManager.enableSafeMode(this)
        reloadConfig()
        runOnUiThread {
            Toast.makeText(this, "EMERGENCY RESET triggered", Toast.LENGTH_LONG).show()
        }
    }

    // ═══════════════════════════════════════════════════════════════
    //  Overlay Management
    // ═══════════════════════════════════════════════════════════════

    private fun createOverlayView() {
        if (windowManager == null || isLsposedModeFromConfig()) return

        touchBlockView = TouchBlockView(this)
        touchBlockView?.setService(this)

        val params = WindowManager.LayoutParams(
            WindowManager.LayoutParams.MATCH_PARENT,
            WindowManager.LayoutParams.MATCH_PARENT,
            WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY,
            WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE or
                    WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL or
                    WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN or
                    WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS,
            PixelFormat.TRANSLUCENT
        )
        params.gravity = Gravity.TOP or Gravity.START
        overlayParams = params

        touchBlockView?.apply {
            setRegions(regions)
            setBlockingEnabled(isEnabled && !forceSafeMode)
            setBlockCount(blockCount.get())
        }

        try {
            windowManager?.addView(touchBlockView, params)
            updateOverlayTouchThroughput()
        } catch (e: Exception) {
            Log.e(TAG, "Failed to add overlay", e)
        }
    }

    private fun updateOverlayTouchThroughput() {
        val view = touchBlockView ?: return
        val params = overlayParams ?: return
        val wm = windowManager ?: return

        val shouldBlockTouches = isEnabled && !forceSafeMode && regions.isNotEmpty() && !paused

        if (shouldBlockTouches) {
            params.flags = params.flags and WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE.inv()
        } else {
            params.flags = params.flags or WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE
        }

        try {
            wm.updateViewLayout(view, params)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to update overlay touch behavior", e)
        }
    }

    private fun updateOverlayState() {
        touchBlockView?.apply {
            setRegions(regions)
            setBlockingEnabled(isEnabled && !forceSafeMode && !paused)
            setBlockCount(blockCount.get())
        }
        updateOverlayTouchThroughput()
        val nm = getSystemService(NotificationManager::class.java)
        nm?.notify(NOTIFICATION_ID, createNotification())
    }

    // ═══════════════════════════════════════════════════════════════
    //  Broadcast Receiver Registration
    // ═══════════════════════════════════════════════════════════════

    private fun registerConfigReceiver() {
        configReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context?, intent: Intent?) {
                when (intent?.action) {
                    "com.inputblocker.RELOAD" -> reloadConfig()
                    "com.inputblocker.DISABLE" -> disableBlocking()
                    "com.inputblocker.ENABLE" -> enableBlocking(intent.getBooleanExtra("force_safe", true))
                    "com.inputblocker.PAUSE" -> pauseBlocking(60000)
                    "com.inputblocker.RESUME" -> resumeBlocking()
                }
            }
        }
        val filter = IntentFilter().apply {
            addAction("com.inputblocker.RELOAD")
            addAction("com.inputblocker.DISABLE")
            addAction("com.inputblocker.ENABLE")
            addAction("com.inputblocker.PAUSE")
            addAction("com.inputblocker.RESUME")
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            registerReceiver(configReceiver, filter, RECEIVER_NOT_EXPORTED)
        } else {
            @Suppress("UnspecifiedRegisterReceiverFlag")
            registerReceiver(configReceiver, filter)
        }
    }

    private fun disableBlocking() {
        isEnabled = false
        runOnUiThread { updateOverlayState() }
    }

    private fun enableBlocking(clearSafeMode: Boolean) {
        if (clearSafeMode) forceSafeMode = false
        isEnabled = true
        runOnUiThread { updateOverlayState() }
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        intent?.action?.let { action ->
            when (action) {
                "RELOAD" -> reloadConfig()
                "DISABLE" -> disableBlocking()
                "ENABLE" -> enableBlocking(true)
            }
        }
        return START_STICKY
    }

    override fun onDestroy() {
        super.onDestroy()
        isRunning = false
        configObserver?.stop()
        configReceiver?.let {
            try { unregisterReceiver(it) } catch (e: Exception) {
                Log.e(TAG, "Error unregistering config receiver", e)
            }
        }
        try { unregisterReceiver(notificationReceiver) } catch (_: Exception) { }
        touchBlockView?.let { windowManager?.removeView(it) }
    }

    @Nullable
    override fun onBind(intent: Intent?): IBinder? = null

    private fun runOnUiThread(action: Runnable) {
        android.os.Handler(android.os.Looper.getMainLooper()).post(action)
    }

    // ═══════════════════════════════════════════════════════════════
    //  TouchBlockView
    // ═══════════════════════════════════════════════════════════════

    inner class TouchBlockView(context: Context) : View(context) {
        private var serviceRef = WeakReference<OverlayService>(null)
        fun setService(service: OverlayService) { serviceRef = WeakReference(service) }

        private val blockPaint = Paint().apply {
            color = Color.parseColor("#1A00FF00"); style = Paint.Style.FILL
        }
        private val borderPaint = Paint().apply {
            color = Color.parseColor("#00FF00"); style = Paint.Style.STROKE; strokeWidth = 4f
        }
        private val textPaint = Paint().apply {
            color = Color.parseColor("#00FF00"); textSize = 36f; isFakeBoldText = true
        }
        private val countPaint = Paint().apply {
            color = Color.argb(180, 0, 255, 0); textSize = 24f
        }

        private val regionsList = mutableListOf<Region>()
        private var enabled = true
        private var count = 0

        fun setRegions(list: List<Region>) {
            regionsList.clear(); regionsList.addAll(list); invalidate()
        }
        fun setBlockingEnabled(e: Boolean) { enabled = e; invalidate() }
        fun setBlockCount(c: Int) { count = c; invalidate() }

        override fun onDraw(canvas: Canvas) {
            super.onDraw(canvas)
            if (!enabled) return

            for (region in regionsList) {
                when (region.type) {
                    0 -> {
                        val rect = android.graphics.RectF(
                            region.x1 * width, region.y1 * height,
                            region.x2 * width, region.y2 * height
                        )
                        canvas.drawRect(rect, blockPaint)
                        canvas.drawRect(rect, borderPaint)
                    }
                    1 -> {
                        val cx = region.x1 * width
                        val cy = region.y1 * height
                        val r = region.x2 * width
                        canvas.drawCircle(cx, cy, r, blockPaint)
                        canvas.drawCircle(cx, cy, r, borderPaint)
                    }
                    2 -> {
                        val rect = android.graphics.RectF(
                            (region.x1 - region.x2) * width,
                            (region.y1 - region.y2) * height,
                            (region.x1 + region.x2) * width,
                            (region.y1 + region.y2) * height
                        )
                        canvas.drawOval(rect, blockPaint)
                        canvas.drawOval(rect, borderPaint)
                    }
                }
            }

            val statusMsg = if (paused) {
                val remaining = ((pauseExpirationTime - System.currentTimeMillis()) / 1000).coerceAtLeast(0)
                "PAUSED — ${remaining}s remaining"
            } else if (regionsList.isEmpty()) {
                "InputBlocker: No zones configured"
            } else {
                "BLOCKING ${regionsList.size} zone(s)"
            }
            canvas.drawText(statusMsg, 10f, 50f, textPaint)
            if (count > 0) {
                canvas.drawText("Total: $count blocked", 10f, 90f, countPaint)
            }
        }

        override fun onTouchEvent(event: MotionEvent): Boolean {
            if (!enabled || regionsList.isEmpty() || paused) return false

            val nx = event.x / width
            val ny = event.y / height

            if (event.action == MotionEvent.ACTION_DOWN || event.action == MotionEvent.ACTION_MOVE) {
                for (region in regionsList) {
                    if (region.isExclude && isPointInRegion(nx, ny, region)) return false
                }

                for (region in regionsList) {
                    if (!region.isExclude && isPointInRegion(nx, ny, region)) {
                        val contactArea = event.pressure
                        val duration = event.eventTime - event.downTime

                        if (contactArea < region.minPressure || duration > region.maxDuration) {
                            blockCount.incrementAndGet()

                            val now = System.currentTimeMillis()
                            if (now - lastBlockLogTime > RATE_LIMIT_MS) {
                                lastBlockLogTime = now
                                val timeStr = SimpleDateFormat(
                                    "HH:mm:ss", Locale.getDefault()
                                ).format(Date())
                                addBlockEntry(
                                    BlockLogActivity.BlockEntry(
                                        timeStr,
                                        "Blocked at (%.2f, %.2f)  count=%d".format(
                                            nx, ny, blockCount.get()
                                        )
                                    )
                                )
                            }
                            postInvalidate()
                            return true
                        }
                    }
                }
            }
            return false
        }

        private fun isPointInRegion(nx: Float, ny: Float, region: Region): Boolean {
            return when (region.type) {
                0 -> nx >= region.x1 && nx <= region.x2 && ny >= region.y1 && ny <= region.y2
                1 -> {
                    val dx = (nx - region.x1) * width
                    val dy = (ny - region.y1) * height
                    val r = region.x2 * width
                    (dx * dx + dy * dy) <= (r * r)
                }
                2 -> {
                    val dx = (nx - region.x1)
                    val dy = (ny - region.y1)
                    val rx = region.x2
                    val ry = region.y2
                    (dx * dx) / (rx * rx) + (dy * dy) / (ry * ry) <= 1.0f
                }
                else -> false
            }
        }
    }
}
